distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso

```diff
@@ -507,6 +510,52 @@ type Table
         new_names = this.columns.map mapper
         this.take_end (this.length - 1) . rename_columns (Column_Mapping.By_Position new_names) on_problems=on_problems warnings=warnings
 
+
+    ## Prototype Group By function
+    group_by : Column_Selector -> [Aggregate_Column] -> Problem_Behavior -> Warnings.Warning_System -> Table
+    group_by selector columns (on_problems=Report_Warning) (warnings=Warnings.default) =
+        # Grouping Key
+        key_columns = if selector.is_nothing then [] else
+            Table_Helpers.select_columns internal_columns=this.columns selector=selector reorder=True on_problems=on_problems warnings=warnings
+        key_length = key_columns.length
+        make_key i = Group_By_Key <| key_columns.map v->(v.at i)
+
+        # New Table Accumulator
+        new_table = (key_columns.map c->c.name)+(columns.map c->(c.column_name this)) . map n->[n, Vector.new_builder]
+        add_row key =
+            idx = new_table.at 0 . at 1 . length
+            0.up_to key_length . each i->
+                new_table.at i . at 1 . append (key.values.at i)
+            0.up_to (columns.length) . each i->
+                column = columns.at i
+                new_table.at (i + key_length) . at 1 . append (column.initial_value)
+            idx
+
+        # Fold
+        aggregators = columns.map c->(c.make_aggregator this)
+        aggregate map i =
+            key = make_key i
+            row_index = map.get_or_else key (add_row key)
+
+            # Accumulate...
+            0.up_to (columns.length) . each j->
+                aggregator = aggregators.at j
+                array = new_table.at (j + key_length) . at 1 . to_array
+                current = array . at row_index
+                new = aggregator current i
+                array . set_at row_index new
```

Comment:
This may stop working when we fix the "`to_array` is leaking mutability" issue.

As this prototype may live shorter than the above issue - probably ignore this comment.

But just for future - maybe we could just create an `Expandable_Array` which wraps a regular, explicitly-mutable array but also has the `append` operation and does the allocations that give us O(1) amortized complexity of `append`. Because I don't like that we are abusing the `Vector_Builder` a bit when we could just create a data structure meant for such use-cases. But since this is a prototype - just commenting a suggestion for future.